sdk/provisioning/Azure.Provisioning/tests/Functions/JoinFunctionTests.cs

```diff
@@ -0,0 +1,61 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using Azure.Provisioning.Expressions;
+using Azure.Provisioning.Primitives;
+using NUnit.Framework;
+
+namespace Azure.Provisioning.Tests.Functions
+{
+    public class JoinFunctionTests
+    {
+        [TestCase(new string[0], ",", ExpectedResult = "join([], ',')")]
+        [TestCase(new string[] { "a" }, ",", ExpectedResult = "join([\r\n  'a'\r\n], ',')")]
+        [TestCase(new string[] { "a", "b" }, ",", ExpectedResult = "join([\r\n  'a'\r\n  'b'\r\n], ',')")]
+        public string JoinFunction_WithBothLiterals(string[] inputArray, string delimiter)
+        {
+            BicepList<string> array = [.. inputArray];
+            return BicepFunction.Join(array, delimiter).ToString();
+        }
+
+        [TestCase("test", ",", ExpectedResult = "join(test.array, ',')")]
+        public string JoinFunction_WithLiteralDelimiter(string resourceName, string delimiter)
+        {
+            var resource = new TestResource(resourceName);
+            return BicepFunction.Join(resource.Array, delimiter).ToString();
+        }
+
+        [TestCase("test", ExpectedResult = "join(test.array, test.delimiter)")]
+        public string JoinFunction_WithBothExpressions(string resourceName)
+        {
+            var resource = new TestResource(resourceName);
+            return BicepFunction.Join(resource.Array, resource.Delimiter).ToString();
```

Comment:
Now our `BicepList<T>` is always treated as "literal array" therefore no matter this `Array` property is output or not, we always get a `[]` in the result.
Should it have the ability to produce an expression?
As a comparison, the `delimiter` property here as an output, it will produce an expression: `test.delimiter`.